### PR TITLE
Join with Unique Index Prefix Fix

### PR DESCRIPTION
**Pull request by maierlars**
### Scope & Purpose

Bugfix for https://github.com/arangodb/arangodb/issues/21243. For a detailed explanation and workaround see [here](https://github.com/arangodb/arangodb/issues/21243#issuecomment-2303966376).

### CODE REVIEW

File: arangod/ClusterEngine/ClusterIndex.cpp [hunk 1]
```diff
@@ -496,5 +496,5 @@ bool ClusterIndex::supportsStreamInterface(
     default:
       break;
   }
-  return false;
+  return {};
```
Comment (mpoeter) [inline arangod/ClusterEngine/ClusterIndex.cpp:499]: I would prefer to be more explicit:
```suggestion
  return Index::StreamSupportResult::makeUnsupported();
```

File: arangod/Aql/Optimizer/Rule/OptimizerRulesJoin.cpp [hunk 2]
```diff
@@ -779,13 +780,17 @@ std::tuple<bool, IndicesOffsets> checkCandidatesEligible(
...
+      if (!supportResult.hasSupport()) {
         allCandidatesSupportStreamInterface = false;
         LOG_JOIN_OPTIMIZER_RULE << "IndexNode's index does not "
                                    "support streaming interface";
         LOG_JOIN_OPTIMIZER_RULE
             << "-> Index name: " << node->getIndexes()[0]->name()
             << ", id: " << node->getIndexes()[0]->id();
+      } else {
+        // the index stream is a unique stream
+        indexOffset.isUniqueStream = supportResult.isUniqueStream();
```
Comment (mpoeter) [inline arangod/Aql/Optimizer/Rule/OptimizerRulesJoin.cpp:793]: The comment is confusing - `isUniqueStream` can be false after all.

File: arangod/RocksDBEngine/RocksDBVPackIndex.cpp [hunk 3]
```diff
@@ -3147,44 +3149,60 @@ std::unique_ptr<AqlIndexStreamIterator> RocksDBVPackIndex::streamForCondition(
...
-  // for persisted indexes, we can only use a prefix of the indexed keys
-  std::size_t idx = 0;
+  // only one key fields is supported currently
   if (streamOpts.usedKeyFields.size() != 1) {
-    return false;
+    return StreamSupportResult::makeUnsupported();
+  }
+  if (streamOpts.usedKeyFields[0] >= fields.size()) {
+    return StreamSupportResult::makeUnsupported();
   }
```
Comment (mpoeter) [inline arangod/RocksDBEngine/RocksDBVPackIndex.cpp:3169]: This condition is new and I don't really get why this is needed?

File: arangod/RocksDBEngine/RocksDBVPackIndex.cpp [hunk 4]
```diff
@@ -3147,44 +3149,60 @@ std::unique_ptr<AqlIndexStreamIterator> RocksDBVPackIndex::streamForCondition(
...
-  return true;
+  if (idx > fields.size()) {
+    return StreamSupportResult::makeUnsupported();
+  }
+
+  // The uniqueness of this index can only be used if all fields are used.
+  // Example: if we index triples [doc.a, doc.b, doc.c] then the uniqueness
+  // refers to the triples and just taking a prefix [doc.a, doc.b] is not
+  // unique. Thus, we should not pick an optimized strategy for joining two
+  // unique indexes.
```
Comment (mpoeter) [inline arangod/RocksDBEngine/RocksDBVPackIndex.cpp:3195]:
```suggestion
  // refers to the triples and just taking a prefix [doc.a, doc.b] is not
  // necessarily unique. Thus, we should not pick an optimized strategy for
  // joining two unique indexes.
```

File: arangod/RocksDBEngine/RocksDBVPackIndex.cpp (hunk 3 above)
Comment (maierlars) [reply arangod/RocksDBEngine/RocksDBVPackIndex.cpp:3169]: It was missing. We can't have a join key on a stored value.

**Review by mpoeter — APPROVED** [review]

**Pull request merged**